### PR TITLE
New data source: azuread_directory_object

### PR DESCRIPTION
**Pull request by Threpio**
https://github.com/hashicorp/terraform-provider-azuread/issues/824

### CODE REVIEW

File: internal/services/directoryobjects/principal_type_data_source.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+	objectId := d.Get("object_id").(string)
+
+	obj, _, err := client.Get(context.Background(), objectId, odata.Query{})
+	if err != nil {
+		return nil
+	}
+
+	// TODO: Not this
+	directoryObject = *obj
```
Comment (Threpio) [inline internal/services/directoryobjects/principal_type_data_source.go:58]: @manicminer - I am pretty sure this is not the way to do it. This is the last thing other than the naming convention to be sorted for this PR/Fix to be good to go I think :)

File: docs/data-sources/principal_type.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+## Example Usage
+
+*Look up and output type of object by ID*
+```terraform
+data "azuread_principal_type" "example" {
+  object_id = "object-id"
+}
+
+output "principal_type" {
+  value = "${data.azuread_principal_type.example.type}"
```
Comment (manicminer) [inline docs/data-sources/principal_type.md:26]: No need for interpolation here
```suggestion
  value = data.azuread_principal_type.example.type
```

File: internal/services/directoryobjects/principal_type_data_source.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+	"github.com/hashicorp/terraform-provider-azuread/internal/tf"
+
+	"github.com/manicminer/hamilton/odata"
+
+	"github.com/hashicorp/terraform-provider-azuread/internal/clients"
+	"github.com/hashicorp/terraform-provider-azuread/internal/validate"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
```
Comment (manicminer) [inline internal/services/directoryobjects/principal_type_data_source.go:17]: Let's tidy up these imports
```suggestion
	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
	"github.com/hashicorp/terraform-provider-azuread/internal/clients"
	"github.com/hashicorp/terraform-provider-azuread/internal/tf"
	"github.com/hashicorp/terraform-provider-azuread/internal/validate"
	"github.com/manicminer/hamilton/msgraph"
	"github.com/manicminer/hamilton/odata"
```

File: docs/data-sources/principal_type.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
+---
+subcategory: "Base"
+---
+
+# Data Source: azuread_principal_type
+
+Gets the OData Object type for a single given Object ID.
```
Comment (manicminer) [inline docs/data-sources/principal_type.md:7]:
```suggestion
Retrieves the OData type for a generic directory object having the provided object ID.
```

File: internal/services/directoryobjects/principal_type_data_source.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+		},
+	}
+}
+
+func principalTypeDataSourceRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client).Users.DirectoryObjectsClient
+	client.BaseClient.DisableRetries = true
+
+	var directoryObject msgraph.DirectoryObject
+
```
Comment (manicminer) [inline internal/services/directoryobjects/principal_type_data_source.go:49]:
```suggestion
	var directoryObject msgraph.DirectoryObject

```

File: internal/services/directoryobjects/principal_type_data_source.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	objectId := d.Get("object_id").(string)
+
+	obj, _, err := client.Get(context.Background(), objectId, odata.Query{})
```
Comment (manicminer) [inline internal/services/directoryobjects/principal_type_data_source.go:52]:
```suggestion
	directoryObject, _, err := client.Get(context.Background(), objectId, odata.Query{})
```

File: internal/services/directoryobjects/principal_type_data_source.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+	if err != nil {
+		return nil
```
Comment (manicminer) [inline internal/services/directoryobjects/principal_type_data_source.go:54]:
```suggestion
		return tf.ErrorDiagPathF(nil, "object_id", "Directory Object with ID %q was not found", objectId)
```

File: internal/services/directoryobjects/principal_type_data_source.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	}
```
Comment (manicminer) [inline internal/services/directoryobjects/principal_type_data_source.go:55]:
```suggestion
	}
	if directoryObject == nil {
		return tf.ErrorDiagF(fmt.Errorf("nil object returned for directory object with ID: %q", objectId), "Bad API Response")
	}
```

File: internal/services/directoryobjects/principal_type_data_source.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+	// TODO: Not this
+	directoryObject = *obj
+
+	d.SetId(*directoryObject.ID)
+
+	switch *directoryObject.ODataType {
+	// There are many more types to switch here - These are the ones added per the issue
```
Comment (manicminer) [inline internal/services/directoryobjects/principal_type_data_source.go:63]:
```suggestion
```

**Review by manicminer — CHANGES_REQUESTED** [review]
Hi @Threpio, thanks for working on this PR. This is looking pretty good! I have made some suggestions inline below, and we will also need to add some acceptance tests for this data source.

If you can take a look at the comments below, and look to add one or two acceptance tests (take a look at some other data sources for inspiration :), then I'll be happy to take another look. Thanks!

**Comment by Threpio** [comment]
Most of the comments have been applied and committed -Sorry for the horrible commit mess.

I am not sure that I have handled the `directoryObject` object properly now that I have changed it to a pointer from the client.Get function.

I will work on the tests asap.

**Comment by manicminer** [comment]
@Threpio Thanks, I'll take another look once the tests are pushed. Don't worry about the commit history, we can squash it :)

**Comment by Threpio** [comment]
Tests have been added -I cannot apply the last suggestion as it is outdated and yet not updated.

For the tests I have done 3 success cases - I do not see examples of failing to terraform configuration being used for different unit tests and so have not written these. The docs are linted but the content might need a 5-minute review.

File: docs/data-sources/directory_object.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+When authenticated with a service principal, this data source requires either `Group.Read.All` or `Directory.Read.All`
+
+When authenticated with a user principal, this data source does not require any additional roles.
+
+## Example Usage
+
+*Look up and output type of object by ID*
+```terraform
+data "azuread_directory_object" "example" {
+  object_id = "object-id"
```
Comment (manicminer) [inline docs/data-sources/directory_object.md:22]: We like to use syntactically appropriate but obviously-fake values in example configs :)
```suggestion
  object_id = "00000000-0000-0000-0000-000000000000"
```

File: docs/data-sources/directory_object.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+output "object_type" {
+  value = data.azuread_directory_object.example.type
+}
+```
+
```
Comment (manicminer) [inline docs/data-sources/directory_object.md:29]: We should include an Arguments section here

```suggestion

## Argument Reference

The following arguments are supported:

* `object_id` - (Optional) Specifies the Object ID of the directory object to look up.

```

File: docs/data-sources/directory_object.md [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+## Attributes Reference 
+
+The following attributes are exported:
+
+*`object_id` - The object_id specified.
```
Comment (manicminer) [inline docs/data-sources/directory_object.md:34]:
```suggestion
*`object_id` - The object ID of the directory object.
```

File: docs/data-sources/directory_object.md [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+*`type` - The type of the object
```
Comment (manicminer) [inline docs/data-sources/directory_object.md:35]:
```suggestion
*`type` - The shortened OData type of the directory object. Possible values include: `Group`, `User` or `ServicePrincipal`.
```

File: docs/data-sources/directory_object.md [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+# Data Source: azuread_directory_object
+
+Retrieves the OData type for a generic directory object having the provided object ID.
+
+## API Permissions
+
+The following API permissions are required in order to use this data source.
+
+When authenticated with a service principal, this data source requires either `Group.Read.All` or `Directory.Read.All`
```
Comment (manicminer) [inline docs/data-sources/directory_object.md:13]:
```suggestion
When authenticated with a service principal, this data source requires either `User.Read.All`, `Group.Read.All` or `Directory.Read.All`, depending on the type of object being queried.
```

**Review by manicminer — APPROVED** [review]
Hi @Threpio, many thanks for working on this and for adding the tests - which should be enough to cover usage of this data source.

I've run through this again and made some changes, mainly renaming the data source from `azuread_principal_type` to `azuread_directory_object` - because I expect we'll support additional object types in future and this will help us avoid deprecating a data source in order to rename it.

In addition to that, I've made some fixes to the tests, added some nil checking that I missed on my first review, and made some documentation updates. I added the latter as inline suggestions below to give you an idea of how we like to structure and word the docs.

This is a great addition, many thanks for your effort on this! As I have made some additional changes, I'm going to open this for another maintainer to review but pending any other changes I'm happy for this to be merged and added to the provider.

**Review by katbyte — APPROVED** [review]
LGTM 🦀

**Pull request merged**